### PR TITLE
Feature/ZCS-11428-NEW

### PR DESCRIPTION
**Pull request by inpati**
This Pull Request is having missing changes from https://github.com/Zimbra/zm-mailbox/pull/1306
These changes were identified while testing

### CODE REVIEW

File: store/src/java/com/zimbra/cs/service/admin/CreateVolume.java [hunk 1]
```diff
@@ -51,35 +51,46 @@ private CreateVolumeResponse handle(CreateVolumeRequest req, Map<String, Object>
...
-                // Update JSON state server properties
+                // create external volume info handler
                 ExternalVolumeInfoHandler extVolInfoHandler = new ExternalVolumeInfoHandler(prov);
-                extVolInfoHandler.addServerProperties(volInfoRequest);
 
-                // Add External volume info to response
-                volInfoResponse.setVolumeExternalInfo(volInfoRequest.getVolumeExternalInfo());
+                // validation
+                String globalS3BucketId = volInfoRequest.getVolumeExternalInfo().getGlobalBucketConfigurationId();
+                if(false == extVolInfoHandler.validateGlbBucketID(globalS3BucketId)) {
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/service/admin/CreateVolume.java:65]: add space after if

File: store/src/java/com/zimbra/cs/service/admin/CreateVolume.java [hunk 2]
```diff
@@ -51,35 +51,46 @@ private CreateVolumeResponse handle(CreateVolumeRequest req, Map<String, Object>
...
-                throw ServiceException.FAILURE("Error while adding server properties", null);
+                throw ServiceException.FAILURE("Error while processing CreateVolumeRequest", e);
             }
         }
 
+        // create volume response
+        Volume volRequest = VolumeManager.getInstance().create(toVolume(volInfoRequest, enumStoreType));
+        VolumeInfo volInfoResponse = volRequest.toJAXB();
+
+        // as id is created once volume is created, update id
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/service/admin/CreateVolume.java:84]: phrasing could be better as 
`update ID with newly created volume ID`

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 3]
```diff
@@ -192,4 +192,35 @@ public void modifyServerProperties(VolumeInfo volInfo) throws ServiceException,
...
         }
     }
+
+    /**
+     * Validate global bucket ID by reading LDAP atrribute
+     * @param globalS3BucketId
+     * @returns true if "global bucket ID" is valid
+     * @throws ServiceException, JSONException
+     */
+    public Boolean validateGlbBucketID(String globalS3BucketId) throws ServiceException, JSONException {
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:202]: could rename method to `validateGlobalBucketID` from `validateGlbBucketID`

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 4]
```diff
@@ -192,4 +192,35 @@ public void modifyServerProperties(VolumeInfo volInfo) throws ServiceException,
...
+        Boolean isValid = false;
+        try {
+            // step 1: Fetch current JSON state object and current JSON state array
+            String globalExternalStoreConfigJson = provisioning.getConfig().getGlobalExternalStoreConfig();
+            JSONObject currentJsonObject = new JSONObject(globalExternalStoreConfigJson);
+            JSONArray currentJsonArray = currentJsonObject.getJSONArray("global/s3BucketConfigurations");
+
+            // step 2: Find "globalBucketUUID" in current JSON array
+            for (int i = 0; i < currentJsonArray.length(); i++) {
+
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:212]: remove empty line

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 5]
```diff
@@ -192,4 +192,35 @@ public void modifyServerProperties(VolumeInfo volInfo) throws ServiceException,
...
+                // step 3: Mark validation as true if "globalBucketUUID" found
+                if (globalS3BucketId.equalsIgnoreCase(currentJsonArray.getJSONObject(i).getString("globalBucketUUID"))) {
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:214]: add constant for globalBucketUUID

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 6]
```diff
@@ -192,4 +192,35 @@ public void modifyServerProperties(VolumeInfo volInfo) throws ServiceException,
...
+     * @returns true if "global bucket ID" is valid
+     * @throws ServiceException, JSONException
+     */
+    public Boolean validateGlbBucketID(String globalS3BucketId) throws ServiceException, JSONException {
+        Boolean isValid = false;
+        try {
+            // step 1: Fetch current JSON state object and current JSON state array
+            String globalExternalStoreConfigJson = provisioning.getConfig().getGlobalExternalStoreConfig();
+            JSONObject currentJsonObject = new JSONObject(globalExternalStoreConfigJson);
+            JSONArray currentJsonArray = currentJsonObject.getJSONArray("global/s3BucketConfigurations");
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:208]: add constant for `global/s3BucketConfigurations`

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 7]
```diff
@@ -192,4 +192,35 @@ public void modifyServerProperties(VolumeInfo volInfo) throws ServiceException,
...
+     * @param globalS3BucketId
+     * @returns true if "global bucket ID" is valid
+     * @throws ServiceException, JSONException
+     */
+    public Boolean validateGlbBucketID(String globalS3BucketId) throws ServiceException, JSONException {
+        Boolean isValid = false;
+        try {
+            // step 1: Fetch current JSON state object and current JSON state array
+            String globalExternalStoreConfigJson = provisioning.getConfig().getGlobalExternalStoreConfig();
+            JSONObject currentJsonObject = new JSONObject(globalExternalStoreConfigJson);
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: `currentJsonObject` & `currentJsonArray` needs having mode meaning full name in regards to feature or the object its holding

File: store/src/java/com/zimbra/cs/service/admin/CreateVolume.java (hunk 1 above)
Comment (inpati) [reply store/src/java/com/zimbra/cs/service/admin/CreateVolume.java:65]: Done

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java (hunk 3 above)
Comment (inpati) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: Done

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java (hunk 5 above)
Comment (inpati) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: We are soon going to have new repo and constants file for common stuff. Will replace all constants when we add new repo

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java (hunk 7 above)
Comment (inpati) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: Done

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 8]
```diff
@@ -192,4 +192,30 @@ public void modifyServerProperties(VolumeInfo volInfo) throws ServiceException,
...
+
+    /**
+     * Validate global bucket ID by reading LDAP atrribute
+     * @param globalS3BucketId
+     * @returns true if "global bucket ID" is valid
+     * @throws ServiceException, JSONException
+     */
+    public Boolean validateGlobalBucketID(String globalS3BucketId) throws ServiceException, JSONException {
+        Boolean isValid = false;
+
```
[the remark this answers is not in the document]
Comment (inpati) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:204]: done

File: store/src/java/com/zimbra/cs/service/admin/CreateVolume.java [hunk 9]
```diff
@@ -51,35 +51,46 @@ private CreateVolumeResponse handle(CreateVolumeRequest req, Map<String, Object>
...
         }
 
+        // create volume response
+        Volume volRequest = VolumeManager.getInstance().create(toVolume(volInfoRequest, enumStoreType));
+        VolumeInfo volInfoResponse = volRequest.toJAXB();
+
+        // update ID with newly created volume ID
+        volInfoRequest.setId(volRequest.getId());
+
+        // add External volume info to response
```
Comment (RakeshAMore) [inline store/src/java/com/zimbra/cs/service/admin/CreateVolume.java:87]: set it only its external volume

Comment (inpati) [reply store/src/java/com/zimbra/cs/service/admin/CreateVolume.java]: Done

File: store/src/java/com/zimbra/cs/service/admin/CreateVolume.java [hunk 10]
```diff
@@ -51,35 +51,45 @@ private CreateVolumeResponse handle(CreateVolumeRequest req, Map<String, Object>
...
-                // Add External volume info to response
+        if (enumStoreType.equals(Volume.StoreType.EXTERNAL)) {
+            try {
+                volInfoRequest.setId(volRequest.getId());
                 volInfoResponse.setVolumeExternalInfo(volInfoRequest.getVolumeExternalInfo());
+                extVolInfoHandler.addServerProperties(volInfoRequest);
             } catch (ServiceException e) {
                 ZimbraLog.store.error("Error while processing CreateVolumeRequest", e);
                 throw e;
             } catch  (JSONException e) {
```
Comment (RakeshAMore) [inline store/src/java/com/zimbra/cs/service/admin/CreateVolume.java:84]: I see one problem here. Consider the case, Volume info is inserted in the table but it fails to insert external info in the JSON object, you are throwing an error but keeping volume stale entry database.

Comment (inpati) [reply store/src/java/com/zimbra/cs/service/admin/CreateVolume.java:84]: We will just throw error for now

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 11]
```diff
@@ -192,4 +192,26 @@ public void modifyServerProperties(VolumeInfo volInfo) throws ServiceException,
...
+    /**
+     * Validate global bucket ID by reading LDAP atrribute
+     * @param globalS3BucketId
+     * @returns true if "global bucket ID" is valid
+     * @throws ServiceException, JSONException
+     */
+    public Boolean validateGlobalBucketID(String globalS3BucketId) throws ServiceException, JSONException {
+        // step 1: Fetch globalS3Config JSON state object and globalS3Config JSON state array
+        String globalExternalStoreConfigJSON = provisioning.getConfig().getGlobalExternalStoreConfig();
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:204]: suggestion for var naming
`globalExternalStoreConfigJSON` => `globalExternalStoreConfig`
`globalS3ConfigJSONObject` => `globalS3Configs`
`globalS3ConfigJSONArray` => `globalS3ConfigList`

File: soap/src/java/com/zimbra/soap/admin/type/VolumeInfo.java [hunk 12]
```diff
@@ -202,12 +197,12 @@ public short getFbits() {
...
 
-    public void setCurrent(boolean value) {
-        current = ZmBoolean.fromBool(value);
+    public void setIsCurrent(boolean value) {
+        isCurrent = value;
     }
 
-    public Boolean isCurrent() {
-        return ZmBoolean.toBool(current);
+    public boolean getIsCurrent() {
```
Comment (RakeshAMore) [inline soap/src/java/com/zimbra/soap/admin/type/VolumeInfo.java:204]: `isCurrent()` is correct method name.

File: store/src/java/com/zimbra/cs/service/util/VolumeConfigUtil.java [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+     * @return
+     * @throws ServiceException
+     */
+    public static void validateCreateVolumeRequest(CreateVolumeRequest request, VolumeInfo volInfoRequest,
+                                                   Volume.StoreType enumStoreType) throws ServiceException {
+        ExternalVolumeInfoHandler extVolInfoHandler = new ExternalVolumeInfoHandler(Provisioning.getInstance());
+
+        // validate store type
+        Short storeType = volInfoRequest.getStoreType();
+        if (1 != storeType.intValue() && 2 != storeType.intValue()) {
```
Comment (RakeshAMore) [inline store/src/java/com/zimbra/cs/service/util/VolumeConfigUtil.java:47]: I think this statement should look like.
`if (StoreType.INTERNAL.getStoreType() != storeType.intValue() && StoreType.EXTERNAL.getStoreType()  != storeType.intValue())`

File: soap/src/java/com/zimbra/soap/admin/type/VolumeInfo.java (hunk 12 above)
Comment (inpati) [reply soap/src/java/com/zimbra/soap/admin/type/VolumeInfo.java:204]: Fixed

File: store/src/java/com/zimbra/cs/service/util/VolumeConfigUtil.java (hunk 13 above)
Comment (inpati) [reply store/src/java/com/zimbra/cs/service/util/VolumeConfigUtil.java:57]: Done

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java (hunk 11 above)
Comment (inpati) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java]: Fixed

File: soap/src/java/com/zimbra/soap/admin/type/VolumeInfo.java [hunk 14]
```diff
@@ -66,7 +65,7 @@ public final class VolumeInfo {
      * @zm-api-field-description Specifies whether blobs in this volume are compressed
      */
     @XmlAttribute(name=AdminConstants.A_VOLUME_COMPRESS_BLOBS /* compressBlobs */, required=false)
-    private ZmBoolean compressBlobs = ZmBoolean.fromBool(false);
```
Comment (RakeshAMore) [inline soap/src/java/com/zimbra/soap/admin/type/VolumeInfo.java:69]: Any benefit of not using `ZmBoolean.fromBool(false);`

**Review by RakeshAMore — APPROVED** [review]

**Review by deepakzimb — APPROVED** [review]

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 15]
```diff
@@ -192,4 +192,48 @@ public void modifyServerProperties(VolumeInfo volInfo) throws ServiceException,
...
+
+    /**
+     * Validate if external volume entry is present in JSON or not
+     * @param volumeId
+     * @returns true if external volume entry is present
+     * @throws ServiceException, JSONException
+     */
+    public Boolean isVolumePresentInJson(int volumeId) throws ServiceException, JSONException {
+      String serverS3ConfigJSON = provisioning.getLocalServer().getServerExternalStoreConfig();
+      JSONObject serverS3ConfigJSONObject = new JSONObject(serverS3ConfigJSON);
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:230]: can we rename serverS3ConfigJSONObject &  serverS3ConfigJSONArray these as well

**Review by dasiyogesh — APPROVED** [review]

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java [hunk 16]
```diff
@@ -192,4 +192,48 @@ public void modifyServerProperties(VolumeInfo volInfo) throws ServiceException,
...
+    }
+
+    /**
+     * Validate if external volume entry is present in JSON or not
+     * @param volumeId
+     * @returns true if external volume entry is present
+     * @throws ServiceException, JSONException
+     */
+    public Boolean isVolumePresentInJson(int volumeId) throws ServiceException, JSONException {
+      String serverS3ConfigJSON = provisioning.getLocalServer().getServerExternalStoreConfig();
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:229]: Indentation is incorrect

Comment (inpati) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:229]: Fixed

File: store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java (hunk 15 above)
Comment (inpati) [reply store/src/java/com/zimbra/util/ExternalVolumeInfoHandler.java:230]: Fixed

**Pull request merged**